README.md
# Инструкция
Exe-файл создается за счет команды линковки четырех файлов
```
gcc ./main.s ./func.s ./get-rand-num-func.s ./itoh.s -o ./solution.exe

```
1. Во время запуска программы пользователю требуется по средствам параметров запуска ввести параметр рандома.
```
./solution.exe 12
```
2. Следующим шагом пользователю требуется ввести длину строки (При использовании тестов записанных в файл, требует указывать размер 80, тк там лежат строки длины 70)
3. Дальнейшее взаимодействие с программой происходит за счет вывода соответствующих сообщений в консоли

[список изменений](https://github.com/zemlianin/ABC-HM-2/blob/master/info/Changes.md)
# Отчет на 4 балла
### Приведено решение задачи на C

1. [main.c](https://github.com/zemlianin/ABC-HM-2/blob/master/main.c)
2. [itoh.c](https://github.com/zemlianin/ABC-HM-2/blob/master/itoh.c)
3. [func.c](https://github.com/zemlianin/ABC-HM-2/blob/master/func.c)
4. [get-rand-num-func.c](https://github.com/zemlianin/ABC-HM-2/blob/master/get-rand-num-func.c)

### Добавлены подробные коментарии в каждый Assebly файл
### Удалены макросы 
Флагами:
```
gcc -masm=intel \
    -fno-asynchronous-unwind-tables \
    -fno-jump-tables \
    -fno-stack-protector \
    -fno-exceptions \
    ./code.c \
    -S -o ./code.s
 ```
 В ручную:
- cdqe
- cdq
- endbr64
### Модифицированная ассемблерная программа отдельно откомпилирована
и скомпонована без использования опций отладки.

### Представлено полное тестовое покрытие, дающее одинаковый результат
Тестирование происходит за счет тестов внутри файла tests.txt, который запускается при выборе нужного режима 0 внутри мэйн

# Отчет на 5 баллов
- Используется функции Task, itoh, Rand, пример:
```
void Task(int *array, int *size, int *B)
void Rand(char *arg, int size, int range, int *A)
char itoh(int a)

```
- Используются локальные переменные i, command, size, input...
- Добавленые коментарии к параметрам функций, примеры:
```
itoh:
mov r13d, edi  # Запсиь n в регистр
```
```
Task:
mov r8, rdi # Запись *array
mov r9, rsi # Запись *size
```
```
main:
mov r14d, edi   # Запись *argc
mov r15, rsi  # Запись *argv

```
- Добавлены коментарии о связи переменных и стека, примеры:
```
lea rdx, -176[rbp]   # Запись адреса массива А в регистр
mov eax, DWORD PTR -4[rbp]  # Запись параметра в регистр
```
```
mov eax, DWORD PTR -180[rbp] # Запись переменной сайз в регистр
cmp DWORD PTR -4[rbp], eax  # Сравнение ее с параметром цикла
```
 
# Отчет на 6 баллов
- Рефакторинг программы на ассемблере за счет максимального использования регистров процессора.

## Main.s:
argc -> r14d

argv -> r15

## func.s:
*array -> r8

*size -> r9

current -> r10d

j -> r11d

code -> r12d

Все переменные отвечающие за работу функции Task были перенесены из стека в регистры

## itoh.s:
i -> r13d

Все переменные отвечающие за работу функции Rand были перенесены из стека в регистры
<br />
- Добавление комментариев, поясняющих эквивалентное использование регистров вместо переменных исходной программы на C.
- Представление результатов тестирования прграммы:

Для сравнения времени работы программ были запущены файлы до оптимизации и после в режиме тестирования, результаты представлены в таблице

<br />

| Номер попытки | До оптимизации, ms  | После оптимизации, ms |
| --- | --- | --- |
| 1 | 69  | 75  |
| 2 | 70  |  77 |
| 3 | 61  | 74  |
| 4 | 76  | 62  |
| 5 | 73  | 38  |
| 6 | 86  | 73  |
| 7 | 96  | 49  |
| 8 | 79  | 45  |
| sum |  610 | 563   |

# Отчет на 7 баллов

- Добавлена функция чтения из файла 
- Добавлена функция вывода в файл
- В файле input содержаться тестовые данные
- В файле output результат работы программы на этих данных

# Отчет на 8 баллов
- В программу добавлен метод Rand, принимающий на вход массив А и заполняющий его случайными числами из требуемого диапозона
- Ввод праметра для генератора при запуске программы
- Добавлена возможность замера времени, вывод происходит в консоль по завершении работы программы, или в случае тестового режима в специально предназначеный файл times.txt

# Отчет на 9 баллов

Для оптимизации кода был задействован флаг -о0
|Параметр|Программа без оптимизации по скорости|Программа с оптимизациями|
|---|---|---|
|Суммарное время работы за 8 запсуков| 603 | 458 |
|Количество строк| 197 | 192 |
|Память|106kb|106kb|


info/Changes.md
# Список изменений
### 1. Удаление ненужных перезаписываний регистров, таких как:

``` 
mov	eax, DWORD PTR -8[rbp]
mov	edi, eax
  ```
  ```
lea	ecx, 97[rax]
mov	eax, DWORD PTR -4[rbp]
movsx	rdx, eax
  ```
  ```
mov	eax, DWORD PTR -4[rbp]
movsx	rdx, eax
```
```
lea	rax, .LC0[rip]
mov	rdi, rax
  ```
  
  ```
  mov	edx, DWORD PTR -8[rbp]
movsx	rdx, edx
  ```
  
  ```
 mov	eax, DWORD PTR -8[rbp]
movsx	rdx, eax
  ```
  
  ```
lea	rax, -184[rbp]
mov	rsi, rax
lea	rax, .LC2[rip]
mov	rdi, rax
  ```
  
  ```
lea	rdx, -188[rbp]
lea	rax, -176[rbp]
mov	rsi, rdx
mov	rdi, rax
  ```
  
  ```
 mov	edx, eax
mov	rax, QWORD PTR -48[rbp]
lea	rcx, .LC2[rip]
mov	rsi, rcx
mov	rdi, rax
  ```
  
  ```
 lea	rax, .LC11[rip]
mov	rsi, rax
  ```
  
  ```
  lea	rax, .LC4[rip]
mov	rsi, rax
lea	rax, .LC5[rip]
mov	rdi, rax
  ```
  
  ```
  mov	rax, QWORD PTR -24[rbp]
lea	rcx, .LC1[rip]
mov	rsi, rcx
mov	rdi, rax
  ```
  
  ```
lea	rax, -184[rbp]
mov	rsi, rax
lea	rax, .LC2[rip]
mov	rdi, rax
```

```
lea	rax, .LC9[rip]
mov	rsi, rax
```

```
lea	rax, .LC12[rip]
mov	rdi, rax
  ```
  
 ### Вместо этого, строка/адрес/переменная записываются напрямую в конечных регистр, без посредника rax или rcx
 ### 2. Удаление кода сгенерированного GCC
 ```
 	.ident	"GCC: (Ubuntu 11.2.0-19ubuntu1) 11.2.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	1f - 0f
	.long	4f - 1f
	.long	5
0:
	.string	"GNU"
1:
	.align 8
	.long	0xc0000002
	.long	3f - 2f
2:
	.long	0x3
3:
	.align 8
4:
```
### 3. Повсеместное удаление макросов cdqe, cdq
### 4. Удаление макроса endbr64
### 5. Оптимизация за счет хранения некторых переменных стека в регистре, среди которых:
## Main.s:
argc -> r14d

argv -> r15

## func.s:
*array -> r8

*size -> r9

current -> r10d

j -> r11d

code -> r12d

Все переменные отвечающие за работу функции Task были перенесены из стека в регистры
## itoh.s:
i -> r13d

Все переменные отвечающие за работу функции Rand были перенесены из стека в регистры



func.c
#include <string.h>

extern char itoh(int a);

void Task(char *array, int *size)
{
    char B[256];
    int j;
    int current = 0;
    for (j = 0; j < *size; ++j)
    {
        if (array[j] == 'a' || array[j] == 'e' || array[j] == 'i' || array[j] == 'u' ||
            array[j] == 'y' || array[j] == 'o' ||
            array[j] == 'A' || array[j] == 'E' ||
            array[j] == 'I' || array[j] == 'U' ||
            array[j] == 'Y' || array[j] == 'O')
        {
            int code = (int)array[j];
            B[current] = '0';
            current++;
            B[current] = 'x';
            current++;
            B[current] = itoh(code / 16);
            current++;
            B[current] = itoh(code % 16);
            current++;
        }
        else
        {
            B[current] = array[j];
            current++;
        }
    }
    *size = current;
    strncpy(array, B, current);
}


func.s

	.intel_syntax noprefix
	.text				# Объявление новой области
	.globl	Task
	.type	Task, @function
Task:
	
	push	rbp			# Начало метода
	mov	rbp, rsp
	sub	rsp, 288
	
	
	mov	r8, rdi	# Запись *array
	mov	r9, rsi	# Запись *size
	
	mov	r10d, 0		# Объявление переменных жи и каррент
	mov	r11d, 0
	
	jmp	.L2				# Прыжок в точку 2
.L6:
	movsx	rdx, r11d		# Запись параметра в регистр
	
	mov	rax, r8	# Расчет текущего элемнета массива
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]		# Сравнение элемента массива с символом гласной буквы
	cmp	al, 97				# Если совпадение, то прыжок в точку 3
	je	.L3
	
	mov	eax, r11d	# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 101
	je	.L3
	
	mov	eax, r11d		# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 105
	je	.L3
	
	mov	eax, r11d		# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 117
	je	.L3
	
	mov	eax, r11d		# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 121
	je	.L3
	
	mov	eax, r11d	# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 111
	je	.L3
	
	mov	eax, r11d		# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 65
	je	.L3
	
	mov	eax, r11d		# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 69
	je	.L3
	
	mov	eax, r11d	# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 73
	je	.L3
	
	mov	eax, r11d	# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 85
	je	.L3
	
	mov	eax, r11d		# Получение элемента массива и сравнение с символом гласной
	movsx	rdx, eax
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 89
	je	.L3
	
	mov	eax, r11d 		# Получение элемента массива и сравнение с символом гласной, 
	movsx	rdx, eax			# если нет совпадения то прыжок на точку 4
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	cmp	al, 79
	jne	.L4

.L3:
	movsx	rdx, r11d		# Получение текущего элемнта массива
	mov	rax, r8
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]		# Преобразование типа
	movsx	eax, al
	mov	r12d, eax		# Запись в переменную код
	
	mov	eax, r10d	# Запись в элемента символа 0
	mov	BYTE PTR -272[rbp+rax], 48
	
	add	r10d, 1		# Карент++
	
	mov	eax, r10d
	mov	BYTE PTR -272[rbp+rax], 120	# Запись в элемент символа х
	
	add	r10d, 1		# Карент++
	
	mov	eax, r12d		# Запись переменной код в регистр
	
	lea	edx, 15[rax]			# Заполнение регистров и вызов метода перевода  системы исчисления
	test	eax, eax			# Деление на 16 и запись в регистры
	cmovs	eax, edx
	sar	eax, 4
	mov	edi, eax
	call	itoh@PLT
	
	movsx	rdx, r10d		# Получение элемента массива с индексом каррент
	mov	BYTE PTR -272[rbp+rdx], al
	
	add	r10d, 1		# Каррент++
	mov	eax, r12d		# Запись элемента массив а в регистр
	
	shr	edx, 28				# Деление на 16 и запись результата в регистры
	add	eax, edx
	and	eax, 15
	sub	eax, edx
	mov	edi, eax
	call	itoh@PLT			# Запуск метода перевод системы исчисления
	
	
	movsx	rdx, r10d		# Получение элемента с индексом каррент
	mov	BYTE PTR -272[rbp+rdx], al	# Запись результата работы функции в элемент массива
	
	add	r10d, 1		# Каррент++
	jmp	.L5
.L4: 
	movsx	rdx, r11d 	
	mov	rax, r8
	add	rax, rdx
	movzx	edx, BYTE PTR [rax]		# Получение элемента ммасива с кординатой жи
	mov	eax, r10d
	mov	BYTE PTR -272[rbp+rax], dl 	# Запись в элемент массива б значения из эррэй
	
	add	r10d, 1		# Карент++
.L5:
	add	r11d, 1		# Жи++
.L2:						
	mov	rax, r9				# Запись адреса сайз в регситр
	mov	eax, DWORD PTR [rax]		
	cmp	r11d, eax		# Сравнение сайз и параметра цикла 
	jl	.L6
	mov	rax, r9
	
	mov	edx, r10d	# Присвоение сайзу каррента
	mov	DWORD PTR [rax], edx
	
	movsx	rdx, r10d	# Запись параметров в регистры
	lea	rcx, -272[rbp]
	mov	rax, r8
	mov	rsi, rcx
	mov	rdi, rax
	
	call	strncpy@PLT			# Вызов метода копирования строк
	
	
	nop					# Выход из метода
	leave
	ret
	.size	Task, .-Task



main.c
#include <time.h>
#include <stdio.h>

extern void Task(char *array, int *size);
extern void Rand(char *arg, int size, int range, char *A);

int main(int argc, char *argv[])
{
	FILE *input, *output;
	char A[100];
	int size, i, j, command, range, sizeN, sumTime = 0;
	char *param_of_random;
	printf("%s", "input size\n");
	scanf("%d", &size);
	printf("%s", "input command: \n 1 - read from file \n 2 - read from console\n 3 - Random\n 0 - Testing");
	scanf("%d", &command);
	if (command == 1)
	{
		input = fopen("files/input.txt", "r");
		for (i = 0; i < size; i++)
		{
			fscanf(input, "%c", &A[i]);
		}
	}
	if (command == 2)
	{
		for (i = 0; i < size; i++)
		{
			scanf("%c", &A[i]);
		}
	}
	if (command == 3)
	{
		range = 26;
		if (argc < 2)
		{
			param_of_random = "0";
		}
		else
		{
			param_of_random = (argv)[1];
		}
		Rand(param_of_random, size, range, A);
	}
	if (command == 0)
	{
		sizeN = size;
		inpu t = fopen("files/tests.txt", "r");
		output = fopen("files/times.txt", "w");
		for (i = 0; i < 50; i++)
		{
			for (j = 0; j < size; j++)
			{
				fscanf(input, "%c", &A[i]);
			}
			clock_t begin = clock();
			Task(A, &sizeN);
			clock_t end = clock();
			fprintf(output, "%d", (int)(end - begin));
			sumTime += (int)(end - begin);
			fprintf(output, "%s", "\n");
		}
		printf("%s", "Sum of Time in ms: ");
		printf("%d", sumTime);
		return 0;
	}
	Task(A, &size);

	output = fopen("files/output.txt", "w");
	printf("%s", "output command: \n 1 - write B to file \n 2 - write B to console\n");
	scanf("%d", &command);

	if (command == 1)
	{
		for (i = 0; i < size; i++)
		{
			fprintf(output, "%c", A[i]);
		}
	}

	if (command == 2)
	{
		for (i = 0; i < size; i++)
		{
			printf("%c", A[i]);
		}
	}
	return 0;
}


get-rand-num-func.s

	.intel_syntax noprefix
	.text				# Новая область
	.globl	Rand			# Объявление функции
	.type	Rand, @function
Rand:
	
	push	rbp			#  Начало метода
	mov	rbp, rsp
	sub	rsp, 48
	
	
	mov	QWORD PTR -24[rbp], rdi	# Параметры метода кладутся в стек
	mov	DWORD PTR -28[rbp], esi
	mov	DWORD PTR -32[rbp], edx
	mov	QWORD PTR -40[rbp], rcx
	
	mov	DWORD PTR -8[rbp], 0	# Присвоение переменной н нуля
	
	mov	rdi, QWORD PTR -24[rbp]  # Вызов метода атои
	call	atoi@PLT
	mov	DWORD PTR -8[rbp], eax	# Получение значения метода и запись в эн
	
	mov	edi, DWORD PTR -8[rbp] 	# Загрузка н в регистр и запуск метода
	call	srand@PLT
	
	mov	DWORD PTR -4[rbp], 0    # Присвоение i = 0
	jmp	.L2			#  Прыжок в точку 2
.L3:
	call	rand@PLT		# Вызов метода ранд

	idiv	DWORD PTR -32[rbp]	# Деление
	mov	eax, edx		
	
	lea	ecx, 97[rax]		# Запись значения чара в регистр
	movsx	rdx, DWORD PTR -4[rbp] # Запись парметра цикла в регистр 
	mov	rax, QWORD PTR -40[rbp]	# Запись Адреса массива в рах
	add	rax, rdx		# Получение кординаты текущего элемнта массива
	mov	edx, ecx
	mov	BYTE PTR [rax], dl	# присвоение эелменту массива нового значения
	add	DWORD PTR -4[rbp], 1 	# Параметр цикла + 1
.L2:
	mov	eax, DWORD PTR -4[rbp]	#  Сравнение параметра цикла с перемнной сайз
	cmp	eax, DWORD PTR -28[rbp] 
	jl	.L3			#  Прыжок в точку 3
	
	
	nop
	nop
	
	
	leave				# Покидание метода
	ret
	.size	Rand, .-Rand
	


main.s
	.intel_syntax noprefix
	.text				# Начало новой секции 
	.section	.rodata		# Объявление строк
.LC0:
	.string	"input size"
.LC1:
	.string	"%d"
	.align 8
.LC2:
	.string	"input command: \n 1 - read from file \n 2 - read from console\n 3 - Random\n 0 - Testing"
.LC3:
	.string	"%s"
.LC4:
	.string	"r"
.LC5:
	.string	"files/input.txt"
.LC6:
	.string	"%c"
.LC7:
	.string	"0"
.LC8:
	.string	"files/tests.txt"
.LC9:
	.string	"w"
.LC10:
	.string	"files/times.txt"
.LC11:
	.string	"Sum of Time in ms: "
.LC12:
	.string	"files/output.txt"
	.align 8
.LC13:
	.string	"output command: \n 1 - write B to file \n 2 - write B to console"
	.text
	.globl	main			# Объявление функции мэйн
	.type	main, @function
main:
	push	rbp			# Начало метода
	mov	rbp, rsp
	sub	rsp, 208
	
	mov	r14d, edi 		# Запись *argc
	mov	r15, rsi		# Запись *argv
	
	mov	DWORD PTR -12[rbp], 0 	# Запись 0 в стек
	
	lea	rax, .LC0[rip]		# Запись параметров метода принт в регистр
	mov	rdi, rax
	call	puts@PLT		# Вызов принт
	
	lea	rsi, -180[rbp]		# Запись параметров метода скан в регистр и вызов метода скан
	lea	rdi, .LC1[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	
	lea	rsi, .LC2[rip]		# Запись параметров метода принт в регистр и вызов принт 
	lea	rdi, .LC3[rip]
	mov	eax, 0
	call	printf@PLT
	
	
	lea	rsi, -184[rbp]
	lea	rdi, .LC1[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT	# Запись параметров метода скан в регистр и вызов метода скан
	
	mov	eax, DWORD PTR -184[rbp]	# Запись комманд в регистр	
	cmp	eax, 1				#  Сравнение комманд с 1
	jne	.L2				#  При неравенстве прыжок в точку 2
	
	lea	rsi, .LC4[rip]			# ЗАпись параметров в регситры
	lea	rdi, .LC5[rip]
	call	fopen@PLT			# Вызов метода открытия потока
	
	mov	QWORD PTR -32[rbp], rax		# Запись результата работы метода в переменную  инпут
	mov	DWORD PTR -4[rbp], 0		# Присвоение параметру цикла 0
	jmp	.L3
.L4:
	lea	rdx, -176[rbp]			# Запись адреса массива А в регистр
	mov	eax, DWORD PTR -4[rbp]		# Запись параметра в регистр
	add	rdx, rax			# Получение значения элеммента с индексом ай
	mov	rdi, QWORD PTR -32[rbp]

	lea	rcx, .LC6[rip]			# Запись параметров метода скан в регистры
	mov	rsi, rcx
	mov	eax, 0
	call	__isoc99_fscanf@PLT		# Вызов метода скан
	
	add	DWORD PTR -4[rbp], 1		# Счетчик цикла++
.L3:
	mov	eax, DWORD PTR -180[rbp]	# Запись переменной сайз в регистр
	cmp	DWORD PTR -4[rbp], eax		# Сравнение ее с параметром цикла
	jl	.L4				# Прыжок в тточку 4
.L2:
	mov	eax, DWORD PTR -184[rbp]	# Запись комманд в регистр
	cmp	eax, 2				# Сравнение комманд с 2
	jne	.L5
	mov	DWORD PTR -4[rbp], 0		# Присвоение переменной ай 0
	jmp	.L6
.L7:
	lea	rdx, -176[rbp]			# Запись адреса массива А в регистр
	mov	eax, DWORD PTR -4[rbp]
	add	rax, rdx			# Получение текущего элемента массива
	mov	rsi, rax			# Запись результата в регистр
	
	
	lea	rax, .LC6[rip]			# Запись параметров метода скан в регситр
	mov	rdi, rax
	mov	eax, 0
	call	__isoc99_scanf@PLT		# Вызов метода скан
	
	add	DWORD PTR -4[rbp], 1		# ай++
.L6:
	mov	eax, DWORD PTR -180[rbp]	# Запись сайз в регистр
	cmp	DWORD PTR -4[rbp], eax		# Сравнение параметра цикла с сайз
	jl	.L7
.L5:
	mov	eax, DWORD PTR -184[rbp]	# Сравнить комманд с 3
	cmp	eax, 3
	jne	.L8
	mov	DWORD PTR -36[rbp], 26		# Рэндж = 26
	cmp	r14d, 1		# Сравнить колличестов переданных в мэйн параметров с 1
	jg	.L9
	lea	rax, .LC7[rip]			# Запись в параметр рандом строчки 0
	mov	QWORD PTR -24[rbp], rax
	jmp	.L10
.L9:
	mov	rax, r15
	mov	rax, QWORD PTR 8[rax]
	mov	QWORD PTR -24[rbp], rax		# Запись в параметр рандома параметр запуска
.L10:
	mov	esi, DWORD PTR -180[rbp]	# ЗАпись параметров метода ранд в регитсры
	lea	rcx, -176[rbp]
	mov	edx, DWORD PTR -36[rbp]
	mov	rdi, QWORD PTR -24[rbp]
	call	Rand@PLT			# Запучк метода ранд
.L8:
	mov	eax, DWORD PTR -184[rbp]	# Запись комманд в регистр и сранение его с 0
	test	eax, eax
	jne	.L11
	
	mov	eax, DWORD PTR -180[rbp]	# СайзН = Сайз
	mov	DWORD PTR -188[rbp], eax	
	
	lea	rsi, .LC4[rip]			#  Запись параметров в регистр и запуск метода открытия потока
	lea	rdi, .LC8[rip]
	call	fopen@PLT			
	
	mov	QWORD PTR -32[rbp], rax		# Запись результата метода открытия потка в переменную инпут
		
	lea	rsi, .LC9[rip]			# Запуск метода открытия потока
	lea	rdi, .LC10[rip]
	call	fopen@PLT
	
	mov	QWORD PTR -48[rbp], rax		#  Запуск метода открытия потока вывода и запись результата в аутпут
	
	mov	DWORD PTR -4[rbp], 0		#  Значение парметра цикла = 0
	jmp	.L12
.L15:	
	mov	DWORD PTR -8[rbp], 0		#  Значение жи = 0
	jmp	.L13
.L14:
	lea	rdx, -176[rbp]
	mov	eax, DWORD PTR -4[rbp]
	add	rdx, rax			#  Расчет значения элемента массива А
	mov	rdi, QWORD PTR -32[rbp]		#  Запись инпут в регистр
	
	lea	rsi, .LC6[rip]			# Запись параметров метода скан в регистр и запуск метода скан
	mov	eax, 0
	call	__isoc99_fscanf@PLT
	
	add	DWORD PTR -8[rbp], 1		# жи++
.L13:
	mov	eax, DWORD PTR -180[rbp]	#  Запись сайз в регистр
	cmp	DWORD PTR -8[rbp], eax		#  Сравнение сайз с жи
	jl	.L14
	
	call	clock@PLT			#  Запуск метода клок
	mov	QWORD PTR -56[rbp], rax		#  Запись результата в бегин
	lea	rsi, -188[rbp]			# Запись параметров в регистры
	lea	rdi, -176[rbp]
	call	Task@PLT			#  Вызов метода таск
	call	clock@PLT			#  Вызов метода клок
	mov	QWORD PTR -64[rbp], rax		#  Запись результата клок в енд
	
	mov	edx, eax			#  Нахождениие разности между енд и бегин
	mov	rax, QWORD PTR -56[rbp]
	mov	ecx, eax
	mov	eax, edx
	sub	eax, ecx
	
	mov	edx, eax			
	
	mov	rdi, QWORD PTR -48[rbp]
	lea	rsi, .LC1[rip]
	mov	eax, 0
	call	fprintf@PLT			# Запуск метода принт
	
	mov	rdx, QWORD PTR -64[rbp]		# Нахождение разности между енд и бегин
	mov	rcx, QWORD PTR -56[rbp]
	sub	rdx, rcx			
	
	add	DWORD PTR -12[rbp], edx		# Прибавить к сумме времен новое время
	
	mov	rsi, QWORD PTR -48[rbp]		# Заполнении регистров и вызов принт	
	mov	edi, 10
	call	fputc@PLT
	
	add	DWORD PTR -4[rbp], 1		# Параметр цикла++
.L12:
	cmp	DWORD PTR -4[rbp], 49		# Сравнение параметра с 50
	jle	.L15
	
	lea	rsi, .LC11[rip]			# Заполнение регистров параметрами
	lea	rdi, .LC3[rip]
	mov	eax, 0
	call	printf@PLT			# Вызов принт
	
	mov	esi, DWORD PTR -12[rbp]		# Заполнение регистров параметрами
	lea	rdi, .LC1[rip]
	mov	eax, 0
	call	printf@PLT			# Вызов принт
	
	mov	eax, 0
	jmp	.L23				# Прыжок в конец программы 
.L11:
	lea	rsi, -180[rbp]			# Запись массива и сайз в регистр
	lea	rdi, -176[rbp]
	call	Task@PLT			# Запуск метода таск
	
	lea	rsi, .LC9[rip]			# Запись параметров в регистры
	lea	rdi, .LC12[rip]			
	call	fopen@PLT			# Открытие потока
	mov	QWORD PTR -48[rbp], rax		# Запись потока в аутпут
	
	lea	rdi, .LC13[rip]			# Запуск метода принт
	call	puts@PLT
	
	lea	rsi, -184[rbp]			#  Запись параметров в регистры
	lea	rdi, .LC1[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT		#  Запуск метода скан
	
	mov	eax, DWORD PTR -184[rbp]	# Сравнение комманд с 1
	cmp	eax, 1
	jne	.L17
	mov	DWORD PTR -4[rbp], 0		#  Присвоить параметру 0
	jmp	.L18
.L19:
	mov	eax, DWORD PTR -4[rbp]
	movzx	eax, BYTE PTR -176[rbp+rax]	# Расчет адреса текущего элемента массива
	movsx	edi, al				# Запись парметров метода принт в регситр
	mov	rsi, QWORD PTR -48[rbp]
	call	fputc@PLT			# Вызов метода принт
	add	DWORD PTR -4[rbp], 1		# Параметр цикла++
.L18:
	mov	eax, DWORD PTR -180[rbp]	# Сравнение сайз с параметром цикла
	cmp	DWORD PTR -4[rbp], eax
	jl	.L19
.L17:
	mov	eax, DWORD PTR -184[rbp]	# Запись комманд в регистр
	cmp	eax, 2				# Сравнение команд с цифрой 2
	jne	.L20
	mov	DWORD PTR -4[rbp], 0		# Обнуление параметра цикла
	jmp	.L21
.L22:
	mov	eax, DWORD PTR -4[rbp]		# Запись параметров в регистры
	movzx	eax, BYTE PTR -176[rbp+rax]
	movsx	edi, al
	call	putchar@PLT			# Вызов метода печати символов
	
	add	DWORD PTR -4[rbp], 1		# Параметр цикла++
.L21:
	mov	eax, DWORD PTR -180[rbp]	# Сравнение параметра цикла с сайз
	cmp	DWORD PTR -4[rbp], eax
	jl	.L22
.L20:
	mov	eax, 0
.L23:
	leave
	ret
	.size	main, .-main
